[BUG]: eigen3 3.4.0 inlining failing on ppc64el architecture: target specific option mismatch

### Required prerequisites

- [X] Make sure you've read the [documentation](https://pybind11.readthedocs.io). Your issue may be addressed there.
- [X] Search the [issue tracker](https://github.com/pybind/pybind11/issues) and [Discussions](https:/pybind/pybind11/discussions) to verify that this hasn't already been reported. +1 or comment there if it has.
- [X] Consider asking first in the [Gitter chat room](https://gitter.im/pybind/Lobby) or in a [Discussion](https:/pybind/pybind11/discussions/new).

### Problem description

When running pybind11 tests, test_eigen.cpp fails to build against the recent new release of eigen3, v3.4.0, on ppc64el architecture.

Test logs can be found at https://ci.debian.net/packages/p/pybind11/unstable/ppc64el/
and https://ci.debian.net/packages/p/pybind11/testing/ppc64el/

The an excerpt of the error message (for pybind11 3.8.1) is
```
[ 28%] Building CXX object CMakeFiles/pybind11_tests.dir/test_docstring_options.cpp.o
[ 31%] Building CXX object CMakeFiles/pybind11_tests.dir/test_eigen.cpp.o
In file included from /usr/include/eigen3/Eigen/Core:286,
                 from /usr/include/pybind11/eigen.h:30,
                 from /tmp/autopkgtest-lxc.b7jzf9a0/downtmp/autopkgtest_tmp/tests-python3.10/test_eigen.cpp:12:
/usr/include/eigen3/Eigen/src/Core/arch/AltiVec/MatrixProductMMA.h: In function ‘Eigen::internal::storeAccumulator<Eigen::internal::blas_data_mapper<double, long, 0, 0, 1>, long, double __vector(2), 2l>(long, long, Eigen::internal::blas_data_mapper<double, long, 0, 0, 1> const&, double __vector(2) const&, __vector_quad*)void’:
/usr/include/eigen3/Eigen/src/Core/util/BlasUtil.h:227:46: error: inlining failed in call to ‘always_inline’ ‘Eigen::internal::blas_data_mapper<double, long, 0, 0, 1>::storePacketBlock<double __vector(2), 4>(long, long, Eigen::internal::PacketBlock<double __vector(2), 4> const&) constvoid’: target specific option mismatch
  227 |   EIGEN_DEVICE_FUNC EIGEN_ALWAYS_INLINE void storePacketBlock(Index i, Index j, const PacketBlock<SubPacket, n> &block) const {
      |                                              ^~~~~~~~~~~~~~~~
In file included from /usr/include/eigen3/Eigen/src/Core/arch/AltiVec/MatrixProduct.h:38,
                 from /usr/include/eigen3/Eigen/Core:350,
                 from /usr/include/pybind11/eigen.h:30,
                 from /tmp/autopkgtest-lxc.b7jzf9a0/downtmp/autopkgtest_tmp/tests-python3.10/test_eigen.cpp:12:
/usr/include/eigen3/Eigen/src/Core/arch/AltiVec/MatrixProductMMA.h:43:44: note: called from here
   43 |   data.template storePacketBlock<Packet, 4>(i, j, tRes);
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
In file included from /usr/include/eigen3/Eigen/Core:350,
                 from /usr/include/pybind11/eigen.h:30,
                 from /tmp/autopkgtest-lxc.b7jzf9a0/downtmp/autopkgtest_tmp/tests-python3.10/test_eigen.cpp:12:
/usr/include/eigen3/Eigen/src/Core/arch/AltiVec/MatrixProduct.h:1137:26: error: inlining failed in call to ‘always_inline’ ‘Eigen::internal::bscale<double __vector(2)>(Eigen::internal::PacketBlock<double __vector(2), 4>&, Eigen::internal::PacketBlock<double __vector(2), 4>&, double __vector(2) const&)void’: target specific option mismatch
 1137 | EIGEN_ALWAYS_INLINE void bscale(PacketBlock<Packet,4>& acc, PacketBlock<Packet,4>& accZ, const Packet& pAlpha)
      |                          ^~~~~~
In file included from /usr/include/eigen3/Eigen/src/Core/arch/AltiVec/MatrixProduct.h:38,
                 from /usr/include/eigen3/Eigen/Core:350,
                 from /usr/include/pybind11/eigen.h:30,
                 from /tmp/autopkgtest-lxc.b7jzf9a0/downtmp/autopkgtest_tmp/tests-python3.10/test_eigen.cpp:12:
/usr/include/eigen3/Eigen/src/Core/arch/AltiVec/MatrixProductMMA.h:41:17: note: called from here
   41 |   bscale<Packet>(tRes, result, alpha);
      |   ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/eigen3/Eigen/Core:350,
                 from /usr/include/pybind11/eigen.h:30,
                 from /tmp/autopkgtest-lxc.b7jzf9a0/downtmp/autopkgtest_tmp/tests-python3.10/test_eigen.cpp:12:
/usr/include/eigen3/Eigen/src/Core/arch/AltiVec/MatrixProduct.h:1199:26: error: inlining failed in call to ‘always_inline’ ‘Eigen::internal::bload<Eigen::internal::blas_data_mapper<double, long, 0, 0, 1>, double __vector(2), long, 2l, 0, 0>(Eigen::internal::PacketBlock<double __vector(2), 4>&, Eigen::internal::blas_data_mapper<double, long, 0, 0, 1> const&, long, long)void’: target specific option mismatch
 1199 | EIGEN_ALWAYS_INLINE void bload(PacketBlock<Packet,4>& acc, const DataMapper& res, Index row, Index col)
      |                          ^~~~~
In file included from /usr/include/eigen3/Eigen/src/Core/arch/AltiVec/MatrixProduct.h:38,
                 from /usr/include/eigen3/Eigen/Core:350,
                 from /usr/include/pybind11/eigen.h:30,
                 from /tmp/autopkgtest-lxc.b7jzf9a0/downtmp/autopkgtest_tmp/tests-python3.10/test_eigen.cpp:12:
/usr/include/eigen3/Eigen/src/Core/arch/AltiVec/MatrixProductMMA.h:39:57: note: called from here
   39 |   bload<DataMapper, Packet, Index, accCols, 0, ColMajor>(tRes, data, i, j);
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
make[2]: *** [CMakeFiles/pybind11_tests.dir/build.make:258: CMakeFiles/pybind11_tests.dir/test_eigen.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:90: CMakeFiles/pybind11_tests.dir/all] Error 2

```
Full log at https://ci.debian.net/data/autopkgtest/unstable/ppc64el/p/pybind11/17286593/log.gz

### Reproducible example code

```text
test_eigen.cpp is located at https://github.com/pybind/pybind11/blob/master/tests/test_eigen.cpp
```

--- a/tools/pybind11Common.cmake
+++ b/tools/pybind11Common.cmake
@@ -318,13 +318,21 @@ function(_pybind11_generate_lto target prefer_thin_lto)
       set(cxx_append ";-fno-fat-lto-objects")
     endif()
 
-    if(CMAKE_CXX_COMPILER_ID MATCHES "Clang" AND prefer_thin_lto)
+    if(CMAKE_SYSTEM_PROCESSOR MATCHES "ppc64le" OR CMAKE_SYSTEM_PROCESSOR MATCHES "mips64")
+      set(NO_FLTO_ARCH TRUE)
+    else()
+      set(NO_FLTO_ARCH FALSE)
+    endif()
+
+    if(CMAKE_CXX_COMPILER_ID MATCHES "Clang"
+       AND prefer_thin_lto
+       AND NOT NO_FLTO_ARCH)
       _pybind11_return_if_cxx_and_linker_flags_work(
         HAS_FLTO_THIN "-flto=thin${cxx_append}" "-flto=thin${linker_append}"
         PYBIND11_LTO_CXX_FLAGS PYBIND11_LTO_LINKER_FLAGS)
     endif()
 
-    if(NOT HAS_FLTO_THIN)
+    if(NOT HAS_FLTO_THIN AND NOT NO_FLTO_ARCH)
       _pybind11_return_if_cxx_and_linker_flags_work(
         HAS_FLTO "-flto${cxx_append}" "-flto${linker_append}" PYBIND11_LTO_CXX_FLAGS
         PYBIND11_LTO_LINKER_FLAGS)

